common/src/main/java/com/microsoft/identity/common/internal/telemetry/Telemetry.java

```diff
@@ -0,0 +1,295 @@
+
+import android.content.Context;
+import android.content.pm.ApplicationInfo;
+import android.content.pm.PackageManager;
+import android.support.annotation.NonNull;
+
+import com.microsoft.identity.common.BuildConfig;
+import com.microsoft.identity.common.internal.logging.DiagnosticContext;
+import com.microsoft.identity.common.internal.logging.Logger;
+import com.microsoft.identity.common.internal.telemetry.adapter.TelemetryAggregationAdapter;
+import com.microsoft.identity.common.internal.telemetry.adapter.TelemetryDefaultAdapter;
+import com.microsoft.identity.common.internal.telemetry.events.BaseEvent;
+import com.microsoft.identity.common.internal.telemetry.observers.ITelemetryAggregatedObserver;
+import com.microsoft.identity.common.internal.telemetry.observers.ITelemetryDefaultObserver;
+import com.microsoft.identity.common.internal.telemetry.observers.ITelemetryObserver;
+import com.microsoft.identity.common.internal.telemetry.rules.TelemetryPiiOiiRules;
+import com.microsoft.identity.common.internal.util.StringUtil;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+import static com.microsoft.identity.common.internal.logging.DiagnosticContext.CORRELATION_ID;
+import static com.microsoft.identity.common.internal.telemetry.TelemetryEventStrings.TELEMETRY_KEY_CORRELATION_ID;
+
+public class Telemetry {
+    private final static String TAG = Telemetry.class.getSimpleName();
+    private static volatile Telemetry sTelemetryInstance = null;
+    private List<ITelemetryObserver> mObservers;
+    private List<Map<String, String>> mTelemetryRawDataMap;
+    private TelemetryConfiguration mDefaultConfiguration;
+    private TelemetryContext mTelemetryContext;
+    private boolean mIsDebugging;
+
+    //if the telemetry configuration is not set
+    private final boolean mIsTelemetryEnabled;
+
+    private Telemetry(final Builder builder) {
+        if (builder == null
+                || builder.mTelemetryContext == null
+                || builder.mDefaultConfiguration == null) {
+            //we do not want to throw exception for telemetry.
+            Logger.warn(
+                    TAG,
+                    "Telemetry is disabled because the Telemetry context or configuration is null"
+            );
+            mIsTelemetryEnabled = false;
+        } else {
+            mIsTelemetryEnabled = true;
+            mDefaultConfiguration = builder.mDefaultConfiguration;
+            mTelemetryContext = builder.mTelemetryContext;
+            mIsDebugging = builder.mIsDebugging;
+            mTelemetryRawDataMap = new LinkedList<>();
+            mObservers = new LinkedList<>();
+        }
+    }
+
+    /**
+     * Prepares instance using builder.
+     **/
+    private static Telemetry prepareInstance(Builder builder) {
+        if (sTelemetryInstance == null) {
+            synchronized (Telemetry.class) {
```

Comment:
Either the null check should move into the `synchronized` block or the whole method should be synchronized. If two threads are competing for this method, synchronizing _after_ the null check is too late